### PR TITLE
add endorsement_repository_inmem realisation with tetst

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: src/storage/endorsements_repository_inmem.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#include "veriblock/storage/endorsements_repository_inmem.hpp"
+
+using namespace VeriBlock;
+
+void EndorsementsRepositoryInmem::put(
+    const VbkPopTx& popTx, const VbkBlock::hash_t& containingBlockHash) {
+  VbkBlock::hash_t endorsedBlockHash = popTx.publishedBlock.getHash();
+  std::string popTxHash = popTx.getHash().asString();
```
Comment (Warchant) [inline src/storage/endorsements_repository_inmem.cpp]: i think you can replace std::string with `decltype(popTx.getHash())`

File: src/storage/endorsements_repository_inmem.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      break;
+    }
+  }
+  if (endrs.empty()) {
+    endorsed_map[endorsedBlockHash] = {{
+        popTx.getHash().asString(),
+        popTx.publishedBlock.getHash(),
+        {containingBlockHash},
+        popTx.blockOfProof.getHash(),
+        popTx.address.toString(),
```
Comment (Warchant) [inline src/storage/endorsements_repository_inmem.cpp]: we store address, but do we use it?

File: src/storage/endorsements_repository_inmem.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  std::vector<Endorsement>& endrs = endorsed_map[endorsedBlockHash];
+  auto endorsement = endrs.begin();
+  for (; endorsement != endrs.end(); ++endorsement) {
+    if (endorsement->endorsementKey == popTxHash) {
+      endorsement->containingVbkHashes.push_back(containingBlockHash);
+      break;
+    }
+  }
+  if (endrs.empty()) {
+    endorsed_map[endorsedBlockHash] = {{
```
Comment (Warchant) [inline src/storage/endorsements_repository_inmem.cpp]: no need to do that. when you execute `endorsed_map[endorsedBlockHash]`, map creates new element and returns it.
so technically you can just push to `endrs`

File: src/storage/endorsements_repository_inmem.cpp (hunk 2 above)
Comment (Mr-Leshiy) [reply src/storage/endorsements_repository_inmem.cpp]: as I remember, no.

Comment (Warchant) [reply src/storage/endorsements_repository_inmem.cpp]: remove it, then

Comment (Mr-Leshiy) [reply src/storage/endorsements_repository_inmem.cpp]: No I have mistaken, we will able to use it in the pop reward algorithm

File: src/storage/endorsements_repository_inmem.cpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+void EndorsementsRepositoryInmem::put(
+    const VbkPopTx& popTx, const VbkBlock::hash_t& containingBlockHash) {
+  VbkBlock::hash_t endorsedBlockHash = popTx.publishedBlock.getHash();
+  std::string popTxHash = popTx.getHash().asString();
+
+  std::vector<Endorsement>& endrs = endorsed_map[endorsedBlockHash];
+  auto endorsement = endrs.begin();
+  for (; endorsement != endrs.end(); ++endorsement) {
+    if (endorsement->endorsementKey == popTxHash) {
+      endorsement->containingVbkHashes.push_back(containingBlockHash);
```
Comment (Warchant) [inline src/storage/endorsements_repository_inmem.cpp]: somewhere here you also need to check that `endorsedVbkHash` also belongs to chain (call func on it)

File: src/storage/endorsements_repository_inmem.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    const VbkBlock::hash_t& endorsedBlockHash,
+    chainContainsBlockFunc func) const {
+  std::vector<Endorsement> result;
+
+  if (endorsed_map.empty()) {
+    return result;
+  }
+
+  for (const auto& endorsement : endorsed_map.at(endorsedBlockHash)) {
+    for (const auto& containingHashes : endorsement.containingVbkHashes) {
```
Comment (Warchant) [inline src/storage/endorsements_repository_inmem.cpp]: somewhere here you also need to check that `endorsedVbkHash` also belongs to chain (call func on it)

**Pull request merged**